binderhub/static/js/components/LinkGenerator.jsx

```diff
@@ -0,0 +1,246 @@
+                onClick={() => setSelectedProvider(p)}
+              >
+                {p.displayName}
+              </a>
+            </li>
+          ))}
+        </ul>
+      </div>
+    </>
+  );
+}
+
+function UrlSelector({ setUrlPath }) {
+  const KINDS = [
+    {
+      id: "file",
+      displayName: "File",
+      placeholder: "eg. index.ipynb",
+      label: "File to open (in JupyterLab)",
+      // Using /doc/tree as that opens documents *and* notebook files
+      getUrlPath: (input) => `/doc/tree/${input}`,
+    },
+    {
+      id: "url",
+      displayName: "URL",
+      placeholder: "eg. /rstudio",
+      label: "URL to open",
+      getUrlPath: (input) => input,
+    },
+  ];
+
+  const [kind, setKind] = useState(KINDS[0]);
+  const [path, setPath] = useState("");
+
+  useEffect(() => {
+    if (path) {
+      setUrlPath(kind.getUrlPath(path));
+    } else {
+      setUrlPath("");
+    }
+  }, [kind, path]);
+  return (
+    <>
+      <label htmlFor="path">{kind.label}</label>
+      <div className="input-group">
+        <input
+          className="form-control"
+          type="text"
+          name="ref"
+          placeholder={kind.placeholder}
+          onChange={(e) => setPath(e.target.value)}
+        />
+        <button
+          className="btn btn-outline-secondary dropdown-toggle"
+          type="button"
+          data-bs-toggle="dropdown"
+          aria-expanded="false"
+        >
+          {kind.displayName}
+        </button>
+        <ul className="dropdown-menu dropdown-menu-end">
+          {KINDS.map((k) => (
+            <li key={k.id}>
+              <a className="dropdown-item" href="#" onClick={() => setKind(k)}>
+                {k.displayName}
+              </a>
+            </li>
+          ))}
+        </ul>
+      </div>
+    </>
+  );
+}
+
+export function LinkGenerator({
+  providers,
+  publicBaseUrl,
+  selectedProvider,
+  setSelectedProvider,
+  repo,
+  setRepo,
+  reference,
+  setReference,
+  urlPath,
+  setUrlPath,
+  isLaunching,
+  setIsLaunching,
+}) {
+  // uh, so react doesn't like it when your prop is named 'ref'
+  const [ref, setRef] = [reference, setReference];
+  const [launchUrl, setLaunchUrl] = useState("");
+
+  useEffect(() => {
+    if (selectedProvider.ref.enabled && ref === "") {
+      setRef(selectedProvider.ref.default);
+    } else {
+      setRef("");
+    }
+  }, [selectedProvider]);
+
+  useEffect(() => {
+    let url;
+    if (repo !== "" && (!selectedProvider.ref.enabled || ref !== "")) {
+      const processedRepo = selectedProvider.repo.preProcess
+        ? selectedProvider.repo.preProcess(repo)
+        : repo;
+      url = new URL(
+        `v2/${selectedProvider.id}/${processedRepo}/${ref}`,
+        publicBaseUrl,
+      );
+      if (urlPath) {
+        url.searchParams.set("urlpath", urlPath);
+      }
+      setLaunchUrl(url.toString());
+    } else {
+      setLaunchUrl("");
+    }
+  }, [repo, ref, urlPath, selectedProvider]);
```

Comment:
@oliverroick ah, right - this is because the rerendering already happens when the props change, and `ref`, `repo`, etc are properties! Is that right?

Reply:
@oliverroick I think I've addressed this. Thanks for the catch!

Reply:
> this is because the rerendering already happens when the props change, and ref, repo, etc are properties! Is that right?

Yes, basically putting this variable in the state and updating in an effect hook that runs on props/state changes will cause an additional re-render. If possible try to avoid effect hooks if you can, and use them only for actual side effects that cannot be handle during the normal component rendering life cycle, like querying an API.